File: core/camera.cc
#include "camera.h"

test camera_test_level = TEST;

camera::camera()
{
    cap[FRONT] = cv::VideoCapture(0, cv::CAP_V4L2); 
    cap[LEFT] = cv::VideoCapture(2, cv::CAP_V4L2);
    cap[LIGHT] = cv::VideoCapture(4, cv::CAP_V4L2);
    cap[BACK] = cv::VideoCapture(6, cv::CAP_V4L2);
    if(!(cap[0].isOpened()&&cap[1].isOpened()&&cap[2].isOpened()&&cap[3].isOpened())) 
        std::cerr << CAMERA_ERROR << std::endl;
    for(int i = 0; i < 4; i++)
    {
        cap[i].set(cv::CAP_PROP_FRAME_WIDTH, 1920);
        cap[i].set(cv::CAP_PROP_FRAME_HEIGHT, 1080);
    }
}

camera::~camera()
{
    for(int i = 0; i < num_camera; i++)
        cap[i].release();
}

error_type camera::do_capture()
{
    std::vector<std::thread> camera_thread;
    std::vector<std::vector<uint8_t>> pictures = std::vector<std::vector<uint8_t>>(num_camera);
    for(int i = 0; i < num_camera; i++)
        //camera_thread.emplace_back(&camera::capture_per_camera,std::ref(cap[i]), std::ref(pictures), static_cast<camera_location>(i));
        camera_thread.emplace_back(std::bind(&camera::capture_per_camera, this, std::ref(cap[i]), std::ref(pictures[i]), static_cast<camera_location>(i)));
    
    for(auto& thread : camera_thread)
        thread.join();

    cap_queue.push(pictures);
    
    if(camera_test_level)
        std::cout << "picture_size" << cap_queue.size() << "\n";

    return DONE;
}

std::vector<std::vector<uint8_t>> camera::pop_cap()
{
    auto tmp = cap_queue.front();
    cap_queue.pop();
    return tmp;
}

error_type camera::capture_per_camera(cv::VideoCapture& video, std::vector<uint8_t>& pictures, const camera_location type)
{
    cv::Mat temp;
    video >> temp;
    if(temp.empty()) return CAMERA_ERROR;

    if (temp.isContinuous()) 
        pictures.assign(temp.data, temp.data + temp.total() * temp.channels());
    else 
    {
        for (int i = 0; i < temp.rows; ++i) 
            pictures.insert(pictures.end(), temp.ptr<uint8_t>(i), temp.ptr<uint8_t>(i) + temp.cols * temp.channels());
    }

    if (camera_test_level)
        test_view(std::ref(temp), type);
    return DONE;
}

error_type camera::test_view(const cv::Mat& pictures, const camera_location type)
{
    std::string filename = "output/frame_" + std::to_string(type) + ".bmp";
    cv::imwrite(filename, pictures);
    std::cout << "Saved " << filename << std::endl;

    return DONE;
}

File: core/camera.py
import cv2
import threading
import os
import time

camera_location = ["front", "", "back"]
class Camera:
    def __init__(self):
        self.cameras = [0, 2]  
        self.cap = [cv2.VideoCapture(cam, cv2.CAP_V4L2) for cam in self.cameras]
        self.fov = 30
        self.lock = threading.Lock()
        if not all([cap.isOpened() for cap in self.cap]):
            print("camera error")
        for cap in self.cap:
            cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1920)  
            self.width = 1280
            cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 1080)
            self.width = 720
        self.is_running = True

        os.makedirs('output', exist_ok=True)
        self.camera_test_level = False
        self.frame_dict = {}

    def __del__(self):
        for cap in self.cap:
            cap.release()
            
    def pop_cap(self):
        det = False
        if self.lock.acquire(timeout=0.5):
            try:
                det = self.frame_dict.copy()
            finally:
                self.lock.release()
        return det
    
    def start_capturing(self):
        self.camera_threads = []
        for idx, cap in enumerate(self.cap):
            thread = threading.Thread(target=self.capture_frames, args=(cap, self.cameras[idx]))
            self.camera_threads.append(thread)
            thread.start()


    def stop_capturing(self):
        for thread in self.camera_threads:
            thread.join()
        self.is_running = False

    def capture_frames(self, video, camera_idx):
        while self.is_running:
            #start_time = time.perf_counter() 
            ret, frame = video.read() 
            if ret:
                if self.lock.acquire(timeout=0.5):
                    try:
                        self.frame_dict[camera_location[camera_idx]] = frame
                    finally:
                        self.lock.release()
                   
                #end_time = time.perf_counter() 
                #capture_duration = (end_time - start_time) * 1000 
                #print(f"Time taken to capture and process frame from camera {camera_idx}: {capture_duration:.2f} ms")

                if self.camera_test_level:
                    self.save_frame(frame, camera_idx)

            else:
                print(f"Failed to capture frame from camera {camera_idx}")

            time.sleep(0.5)

    def save_frame(self, frame, camera_idx):
        filename = f"output/frame_{camera_idx}_{int(time.time() * 1000)}.jpg"
        cv2.imwrite(filename, frame)
        #print(f"Saved {filename}")


if __name__ == "__main__":
    cam = Camera()
    try:
        cam.start_capturing()
    except KeyboardInterrupt:
        cam.stop_capturing()
    finally:
        print(cam.frame_dict)
